### ISSUE TITLE
vertically align = and -> in pattern matching

### ISSUE THREAD

**Issue by enolan** [commercialhaskell/hindent#324]
I'm not sure if this is intentional. This code:

``` haskell
ex1 :: Bool -> Int
ex1 True  = 5
ex1 False = 3

ex2 :: Bool -> Int
ex2 x = case x of
  True  -> 5
  False -> 3
```

Gets reformatted to:

``` haskell
ex1 :: Bool -> Int
ex1 True = 5
ex1 False = 3

ex2 :: Bool -> Int
ex2 x =
  case x of
    True -> 5
    False -> 3
```

In Johan's guide it [says](https://github.com/tibbe/haskell-style-guide/blob/master/haskell-style.md#case-expressions) "Align the `->` arrows when it helps readability." At least in my experience the majority of Haskell is written with vertically aligned pattern matches. The variables are usually aligned too.

**Comment by tfausak** [comment]
I'm not a fan of this style. As soon as one branch changes, every other line has to change to fix the alignment.

**Comment by wiz** [comment]
This can be difficult to get right. But at least hindent can preseve v-align where it is already in place.

**Comment by IvanMalison** [comment]
@tfausak Yeah but since hindent would (hypothetically) handle this automatically, why does that matter?

Are you worried about source control blame/diff sizes? Users who DON'T use hindent?

**Comment by tfausak** [comment]
@IvanMalison I am worried about both of those things. Also the heuristics for aligning things can be difficult. For example, should this be aligned?

``` hs
f Short                  = 0
f AReallyLongConstructor = 1
```

I don't think that is easier to read.

**Comment by wiz** [comment]
```haskell
f Short = 0
f AReallyLongConstructor = 1
f MerelyLong = 2
f AnotherPrettyCons = 3
```

On the other hand, I find this more difficult to follow values while scanning...

**Comment by calvertvl** [comment]
TL;DR: I summarize the existing discussion with some elaboration of my own, then suggest adding an option named ``align-maximum`` to define a tolerance used to determine whether v-align is applied or not:
```` yaml
align-maximum: -1 # current behavior (no v-align)
align-maximum: 0 # always v-align
align-maximum: N # only v-align if it is more than N characters from aligned
````
-----
@enolan, @tfausak, and @wiz have raised some very good points which I have summarized below, with some added comments (as sub-bullets):
* Vertical alignment is very common in available Haskell code.
    * Arguably v-align should be the default in hindent, or users will see odd output and are more likely to decide that hindent is mostly useless.
* Adding/changing a branch can break alignment, causing whitespace changes on multiple lines, and multi-line whitespace changes cause artifacts in diffs and blame.
    * V-alignment is something maintainers _may_ want to avoid, and it is possible that not being able to do so in hindent would cause them to not use a code formatter at all, which would also prevent their contributors from doing so.
* There is apparently some concern about users who do _not_ use hindent.
    * No reasoning was supplied; I can't come up with a convincing reason to worry about this beyond an "opinionated default = lost users" argument.
* Un-aligned is often more difficult to read, but for some users/circumstances aligned is more difficult to read.
    * IMO difficult-to-read aligned code usually means you need to refactor or your names/patterns are too long; I find v-aligned code to be much easier to read in nearly all cases.
* An heuristic to detect and preserve v-align was suggested, but it was noted that this can be difficult.
    * I would argue that for stability, clarity, and speed, we should prefer a formatting operation that can be done based purely on configuration and the AST without any need for separate parsing of the source file to detect pre-existing formatting.

An additional reason of my own:
* If a project's style guidelines are compatible with hindent, then I am more likely to contribute code to that project even if I hate the indentation style used (assuming that hindent supports my own preferred indentation style).

Normally I would argue against adding another option as I generally prefer sensible defaults, and option proliferation IMO tends to result in insensible defaults, but in this case I think it is warranted.

I suggest adding an option, allowing for the flexibility to choose between:
1. Align ``=`` and ``->`` if they would be more than N apart
    * I would prefer a default of N=0 (forced v-align), though 2 might be a reasonable compromise; I generally find differences of more than 3-4 to be significantly more difficult to follow.
1. Never align
    * Preserving this (current behavior) so maintainers who wish can use an un-aligned format does have value.

Suggested option name and values/comments:
```` yaml
align-maximum: 0   # always align, suggested default matching apparent community standard
align-maximum: -1  # preserves current behavior - disable v-align or use (line-length - 1)
align-maximum: N # configurable such that align-maximum is the allowed difference
    # in horizontal position, and exceeding this value forces vertical alignment
````

**Comment by calvertvl** [comment]
Another option I have thought of (remember that pattern matches / guards / cases usually  cannot be re-ordered):

Ragged-right patterns => force v-align:
```` haskell
f Short                  = 0
f AReallyLongConstructor = 1
f MerelyLong             = 2
f AnotherPrettyCons      = 3
````
Pattern length is strictly increasing => don't do any alignment:
```` haskell
g Short = 0
g MerelyLong = 1
g AnotherPrettyCons = 2
g AReallyLongConstructor = 3
````
Pattern length is strictly decreasing => don't do any alignment:
```` haskell
h AReallyLongConstructor = 0
h AnotherPrettyCons = 1
h MerelyLong = 2
h Short = 3
````

**Comment by wiz** [comment]
> Another option I have thought of (remember that pattern matches / guards / cases usually cannot be re-ordered):

While that's a nice catch, it has a major flaw: when you throw in one more long case the whole thing explodes.

In my code style, the overarching rule is "minimize conditionals" - "if"s and "when"s in guidelines not only complicate parser/printer, but drain mental energy too.

**Comment by chrisdone** [comment]
It's not consistent with the rule of always generating code that doesn't depend on name length.

I think a feature to align isn't so bad if enabled by a flag. But considering the code increase in complexity it's not sounding like a great addition to the codebase.

**Comment by calvertvl** [comment]
Always generating output that doesn't depend on name length is a very good goal, but it is not practical as a rule: What if someone uses 40-character names for everything but limits lines to 80 characters?

I would also argue that this is one of the cases where the length of a name would always have an impact, as the effect of having the right hand side on the same line is to shorten the effective line length.

The primary additional complexity in my suggested solution above was because it added an heuristic to sometimes skip the code; arguably that is more in line with Johan Tibell's style guide, but I don't think it is necessary (and I prefer the always-align option anyway).

Simplifying to the enable/disable case removes the heuristic, but the basic algorithm when applying alignment should be the same in any case.

My question: Is some form of alignment going to be implemented, or should I start looking elsewhere for Haskell code formatting?

**Comment by chrisdone** [comment]
I'd definitely look elsewhere to consider all your options in any case. I
don't think I'll be implementing alignment. It's a rule that keeps things
simple.
